[CMake] Replace early swift-syntax with FetchContent

rintaro opened the pull request:
Use FetchContent to include swift-syntax directly in swift. This can be thought of as an add_subdirectory for a directory outside the root.

The default build directory will be _deps/swiftsyntax-subbuild/, though the modules and shared libraries will be built in lib/swift/host by passing down SWIFT_HOST_LIBRARIES_DEST_DIR to avoid copying them as we were doing previously.

bnbarham committed:
[CMake] Replace early swift-syntax with FetchContent

Use FetchContent to include swift-syntax directly in swift. This can be
thought of as an `add_subdirectory` for a directory outside the root.

The default build directory will be `_deps/swiftsyntax-subbuild/`, though
the modules and shared libraries will be built in `lib/swift/host` by
passing down `SWIFT_HOST_LIBRARIES_DEST_DIR` to avoid copying them as we
were doing previously.

rintaro committed:
[CMake] Centralize the logic to get host '-target' tripple

And pass the value to 'swift-syntax', so they gets correct deployment
target.

rintaro committed:
[CMake] Remove SetRPATH.cmake

This used to overwrite the RUNPATH when copying libraries from
earlyswiftsyntax directory. Since earlyswiftsyntax is now replaced with
FetchContent, we don't need this anymore

rintaro committed:
[CMake] update for swift-syntax change

rintaro committed:
[CMake] Set correct RPATH for swift-syntax libs in Linux

rintaro commented:
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please test

rintaro commented:
@swift-ci Please Build Toolchain

rintaro commented:
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please Build Toolchain

rintaro committed:
[CMake] Install swift-syntax libraries

rintaro commented:
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please Build Toolchain

rintaro commented:
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please smoke test

finagolfin commented:
Spurred by #68401, I started looking into cross-compiling swift-syntax for the native Android compiler, only to find early-swift-syntax doesn't support cross-compilation. With a tweak or two to this pull, I think it can be made to work: is that the plan?

Because I don't know about macOS, but the recent shipping of swift-syntax with the linux Swift compiler means that cross-compiling the toolchain is currently broken on Unix platforms, ie swift-syntax won't be cross-compiled.

Also, any plans to backport this pull to 5.9?

rintaro committed:
[CMake] Update swift-syntax libraries installation

rintaro commented:
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please Build Toolchain

rintaro commented:
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please smoke test

rintaro committed:
[CMake] Introduce 'swift-syntax-lib' install component

So that swift-syntax host libraries can be installed independently.

rintaro commented:
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please smoke test

rintaro committed:
[build-script] Add back --install-swiftsyntax

'--install-swift-syntax' is now a sugar of '--install-swift
--swift-install-component=swift-syntax-lib'

rintaro commented:
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please Build Toolchain

rintaro commented:
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please Build Toolchain

rintaro commented:
@finagolfin 
> Spurred by https://github.com/apple/swift/pull/68401, I started looking into cross-compiling swift-syntax for the native Android compiler, only to find early-swift-syntax doesn't support cross-compilation. With a tweak or two to this pull, I think it can be made to work: is that the plan?

The _main_ motivation of this PR is that single `cmake --build /path/to/swift-build-dir`  invocation rebuilds swift-syntax libraries if modified. But I'm aware of that the `earlyswiftsyntax` didn't support cross compiling, and this PR may make it easier. I'd appreciate it if you can check (and fix) it after we merge this PR.

> Also, any plans to backport this pull to 5.9?

I cannot promise anything, but I personally want this in 5.9 too.

bnbarham commented on cmake/modules/AddPureSwift.cmake:
```diff
@@ -46,20 +46,7 @@ function(_add_host_swift_compile_options name)
-    MACCATALYST_BUILD_FLAVOR ""
-    DEPLOYMENT_VERSION "${DEPLOYMENT_VERSION}")
-
+  swift_get_host_triple(target)
```
Can this just use `SWIFT_HOST_TRIPLE`?

rintaro replied:
Updated

finagolfin commented:
> But I'm aware of that the earlyswiftsyntax didn't support cross compiling, and this PR may make it easier. I'd appreciate it if you can check (and fix) it after we merge this PR.

I got the flags added locally to cross-compile this and the regex parser, will submit that once this is in.

I've also backported this to 5.9, along with #68041. Since that adds more checks of `SWIFT_SWIFT_PARSER`, whichever one of these two pulls goes in first will mean the second will have to be updated.

rintaro committed:
[CMake] Adjust FetchContent of 'swift-syntax'

* Move the logic to right before include(lib) because swift-syntax is a
  "lib"
* Use function to limit the scope of variables for FetchContent
* Use standard 'CMAKE_*_OUTPUT_DIRECTORY' instead of custom
  variable names.

rintaro committed:
[CMake] Use global SWIFT_HOST_TRIPLE consistently

Instead of calling 'get_swift_host_triple()' repeatedly, because it
always return the same value.

rintaro committed:
[CMake] Improve argument list handling of force_target_link_libraries

The argument list handling was a hack and confusing.

rintaro commented:
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please Build Toolchain

rintaro commented:
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please test

rintaro committed:
[CMake] Set 'CMAKE_Swift_COMPILER_TARGET' for swift-syntax

bnbarham reviewed (approved):
🥳

rintaro commented:
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please test

rintaro commented:
https://github.com/apple/llvm-project/pull/7424
https://github.com/apple/swift-syntax/pull/2173
@swift-ci Please Build Toolchain
